### PR TITLE
Merge and use time trace sources into dmd

### PR DESCRIPTION
**Pull request by maxhaton**
For anyone who doesn't know this is the time-trace facility from LDC. It's effectively a one-stop-shop for profiling compile times, simultaneously providing a simple high level view (with an appropriate profile viewer) of compilation (at the level of files and imports of those files)

The timetrace.d file is in `common` so it can eventually be used in the backend (code generation, especially as a symptom of overzealous use of templates, is not free and object emission can be very not free since the dmd implementation of (say) ELF doesn't scale to large *N* very well). At the moment however it's only used by the frontend.

@JohanEngelen I had to remove some phobos usage, hopefully i didn't alter anything Also @thewilsonator @kinke to make you aware of an upcoming change if this gets merged.

### CODE REVIEW

**Comment by maxhaton** [comment]
This isn't completely ready yet but I wanna run it through the CI pipelines

File: src/dmd/globals.d [hunk 1]
```diff
@@ -261,6 +261,10 @@ extern (C++) struct Param
     const(char)[] resfile;
     const(char)[] exefile;
     const(char)[] mapfile;
+
+    bool timeTrace;
+    uint timeTraceGranularity = 500;
```
Comment (thewilsonator) [inline src/dmd/globals.d:266]: a comment as to the units  of `timeTraceGranularity` would be good here. Is this microseconds or milliseconds or something else?

Comment (maxhaton) [reply src/dmd/globals.d:266]: noted. I need to update the manual too presumably. it's 4am now so that's something for tomorrow.

**Comment by maxhaton** [comment]
`dmd.globals` thing is going to be annoying. that being said within a reasonable definition of common I guess the globals should be common to them both?

**Comment by dkorpel** [comment]
As a user of `--ftime-trace` in LDC, I like having the same in dmd, especially with the same command line interface / output format. 

@kinke when this gets merged, can you use the same source as DMD or will LDC keep its own time trace implementation? In the latter case, the `version (LDC)` branches in this PR are unnecessary.

> This isn't completely ready yet but I wanna run it through the CI pipelines

I've marked it as draft, please notify when it's ready for review.

File: src/dmd/semantic2.d [hunk 2]
```diff
@@ -75,8 +75,10 @@ enum LOG = false;
  */
 extern(C++) void semantic2(Dsymbol dsym, Scope* sc)
 {
+    import dmd.timetrace_sema;
     scope v = new Semantic2Visitor(sc);
-    dsym.accept(v);
+    scope vtimetrace = new SemanticTimeTraceVisitor!Semantic2Visitor(v);
+    dsym.accept(vtimetrace);
```
Comment (JohanEngelen) [inline src/dmd/semantic2.d:81]: This design was chosen to be able to non-invasively add tracing of the frontend. Now that tracing becomes part of the frontend, a different design choice could be made (e.g. add tracing code directly in SemanticXVisitor), perhaps later.

File: src/dmd/timetrace_sema.d [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    static if (i == "DsymbolSemanticVisitor")
+        private enum pretext = "Sema1: ";
+    else static if (i == "Semantic2Visitor")
+        private enum pretext = "Sema2: ";
+    else static if (i == "Semantic3Visitor")
+        private enum pretext = "Sema3: ";
+    else
+        static assert (false, "did not recognize SemaVisitor ==" ~ SemaVisitor);
+}
+/// Time tracing visitor for semantic analysis. Only valid to instantiate and use this visitor if time tracing is enabled.
```
Comment (JohanEngelen) [inline src/dmd/timetrace_sema.d:83]: It is valid to instantiate when time tracing is not enabled. It's exactly what the code does.

File: src/dmd/timetrace_sema.d [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+extern(C++) final class SemanticTimeTraceVisitor(SemaVisitor) : Visitor
```
Comment (JohanEngelen) [inline src/dmd/timetrace_sema.d:84]: Repeated: This design was chosen to be able to non-invasively add tracing of the frontend. Now that tracing becomes part of the frontend, a different design choice could be made (e.g. add tracing code directly in SemanticXVisitor), perhaps later.
I consider this a little bit as a hack.

File: src/dmd/common/timetrace.d [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+        writeDurationEvents(buf);
+        // Remove the trailing comma (and newline!) to obtain valid JSON.
+        if ((*buf)[buf.length()-2] == ',')
+        {
+            buf.setsize(buf.length()-2);
+            buf.writeByte('\n');
+        }
+    }
+
+    void writeMetadataEvents(OutBuffer* buf)
```
Comment (JohanEngelen) [inline src/dmd/common/timetrace.d:271]: for clarity, change name to `writeMetadata`

File: src/dmd/common/timetrace.d [hunk 6]
```diff
@@ -0,0 +1,438 @@
+/**
+
+Compilation time tracing, --ftime-trace.
+The time trace profile is output in the Chrome Trace Event Format, described
+here: https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview
```
Comment (JohanEngelen) [inline src/dmd/common/timetrace.d:5]: Can you add a comment here that the output is "optimized" for viewing with Tracy, and it should stay like that. Chrome's viewer is essentially unusable. https://github.com/wolfpld/tracy

**Comment by JohanEngelen** [comment]
> @JohanEngelen I had to remove some phobos usage, hopefully i didn't alter anything.

Add a comment to the copied function(s) with a link to phobos. Such that bugfixes in Phobos may at some point also land in the copied functions. (fwiw: I am against the DMD design choice to not use the Dlang standard library).

**Comment by maxhaton** [comment]
> > @JohanEngelen I had to remove some phobos usage, hopefully i didn't alter anything.
> 
> Add a comment to the copied function(s) with a link to phobos. Such that bugfixes in Phobos may at some point also land in the copied functions. (fwiw: I am against the DMD design choice to not use the Dlang standard library).

It was only string manipulation, just pointed it out in case someone is parsing it somewhere (for example).

File: src/dmd/common/timetrace.d [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+        else
+        {
+            filename = "out.time-trace";
+        }
+    }
+    return filename;
+}
+
+extern (C++) void fatal();
+extern (C++) void error(const ref Loc loc, const(char)* format, ...);
```
[the remark this answers is not in the document]
Comment (maxhaton) [reply src/dmd/common/timetrace.d:81]: errors imports globals which currently isn't linked with dmd.common. It seems reasonable to just change that though

File: src/dmd/globals.d (hunk 1 above)
Comment (JohanEngelen) [reply src/dmd/globals.d:266]: I recommend to add the unit as postfix to the name of the variable.

**Comment by RazvanN7** [comment]
@maxhaton Any progress on this?

**Comment by nordlow** [comment]
Does this include memory tracing?

**Comment by JohanEngelen** [comment]
> Does this work include tracing of memory usage?

Yes. Use LDC ;-)

**Comment by nordlow** [comment]
Does this solution include template instance types into the trace?

**Comment by nordlow** [comment]
> Does this solution include template instance types into the trace?

Are you saying memory tracing is not included in this DMD PR?

**Comment by JohanEngelen** [comment]
What I am saying is: just use LDC. It has this functionality for a while now, and you can see for yourself what it does and does not do.

**Comment by nordlow** [comment]
I don't understand how this works. How are probes performed? By stopping the process?

**Comment by WalterBright** [comment]
Could you please provide a link to the documentation on what this switch does?

**Comment by dkorpel** [comment]
There's no written documentation yet, but I created a video tutorial:
https://forum.dlang.org/post/vfnettfnbxblhtldgffp@forum.dlang.org

**Comment by dkorpel** [comment]
Rebooted in https://github.com/dlang/dmd/pull/16363

**Comment by maxhaton** [comment]
Nice. I don't really have time for much DMD stuff these days but I'll try
and help you get that reviewed and merged @dkorpel

**Pull request closed without merging**